### PR TITLE
introduce Kotlinx Benchmark tests (copied from #763)

### PR DESCRIPTION
**Pull request by aSemy**
copied from #763 - credit to @tmszdmsk!

I thought this was a good change, and I'm curious about investigating #13 and I think these benchmark tests can help.

I also refactored the existing 'client tests' to be in a new 'test modules' directory, because I think it's neater and more organised to have the test projects together, rather than mixed amongst the 'regular' projects.

### CODE REVIEW

File: test-modules/performance-tests/README.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+to build the `.jar` file containing the benchmark code. It will be saved to `./build/benchmarks/main/jars` directory.
+
+From there you can run JMH with all the JMH-specific configuration options form CLI (even on different machine).
+
+```shell
+java -jar ./build/benchmarks/main/jars/performance-tests-main-jmh-<version>.jar -w 1 -i 2 -r 60s -prof jfr
+```
+
+which will attach the JFR profiler (not possible with `kontlinx-benchmark` executed from Gradle)
```
Comment (Raibaz) [inline test-modules/performance-tests/README.md:31]: Typo in `kontlinx-benchmark` :)

**Review by Raibaz — COMMENTED** [review]
This looks pretty interesting, thanks for putting it together!

**Comment by aSemy** [comment]
I added some additional tests for context, and used https://jmh.morethan.io/ to visualise the results. 

image


I don't think it's a surprise that the test without mocking/stubbing is very fast, but as commented [here](https://github.com/mockk/mockk/issues/13#issuecomment-990576933) the stub decreases the operations, even though the stub isn't called.

**Comment by Raibaz** [comment]
Yep makes sense, as discussed in #13 the most expensive part is actually creating stubs rather than using them.

**Comment by aSemy** [comment]
I think this PR can be merged in as-is. It's a good basis for future work, be it making the code better, or adding a GHA or visualizer plugin https://github.com/jzillmann/gradle-jmh-report

**Comment by Raibaz** [comment]
Yep I think the next steps about this would be adding a GHA to monitor performances and track potential hits.

Thanks a lot!

**Pull request merged**